File: app/src/main/java/wiki/depasquale/mcachepreview/Github.java
package wiki.depasquale.mcachepreview;

import android.annotation.SuppressLint;
import android.support.annotation.NonNull;
import android.util.Log;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import java.util.concurrent.TimeUnit;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;
import retrofit2.http.GET;
import retrofit2.http.Path;
import wiki.depasquale.mcache.adapters.FilesIOHandler;
import wiki.depasquale.mcache.core.IOHandler;
import wiki.depasquale.mcache.core.MCacheBuilder;

/**
 * diareuse on 26.03.2017
 */

class Github {

  private static Retrofit retrofit;
  private static Service service;

  public static void user(Class<? extends IOHandler> cls) {}

  @SuppressLint("LogConditional") @NonNull
  public static Observable<User> user(String username) {
    return MCacheBuilder
        .request(User.class)
        .using(FilesIOHandler.class)
        .returnCache(true)
        .force(true)
        .descriptor(username)
        .with(getRetrofit()
            .user(username)
            .subscribeOn(Schedulers.newThread())
            .observeOn(AndroidSchedulers.mainThread()));
  }

  private static Service getRetrofit() {
    if (retrofit == null) {
      retrofit = new Retrofit.Builder()
          .baseUrl("https://api.github.com/")
          .client(client())
          .addConverterFactory(GsonConverterFactory.create())
          .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
          .build();
    }
    if (service == null) {
      service = retrofit.create(Service.class);
    }
    return service;
  }

  private static OkHttpClient client() {
    OkHttpClient.Builder client = new OkHttpClient.Builder();
    client.connectTimeout(200, TimeUnit.SECONDS);
    client.readTimeout(200, TimeUnit.SECONDS);
    client.interceptors().add(chain -> {
      Request request = chain.request();
      okhttp3.Response response = null;
      response = chain.proceed(request);
      if (response != null) {
        Log.i("ServiceInfo", response.request().url().url().toString());
      }
      return response;
    });
    return client.build();
  }

  private interface Service {

    @GET("/users/{user}")
    Observable<User> user(@Path("user") String user);
  }

}

File: app/src/main/java/wiki/depasquale/mcachepreview/Github.kt
package wiki.depasquale.mcachepreview

import android.annotation.SuppressLint
import android.util.Log
import com.squareup.okhttp.Interceptor
import com.squareup.okhttp.OkHttpClient
import com.squareup.okhttp.Response
import retrofit.RestAdapter
import retrofit.client.OkClient
import retrofit.http.GET
import retrofit.http.Path
import rx.Observable
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import wiki.depasquale.mcache.Cache
import java.util.concurrent.TimeUnit

/**
 * diareuse on 26.03.2017
 */

object Github {

    private var retrofit: RestAdapter? = null
    private var service: Service? = null

    @SuppressLint("LogConditional")
    fun user(username: String): Observable<User> {
        return Cache.obtain(User::class.java)
            .ofIndex(username)
            .build()
            .getLaterConcat(
                getRetrofit()
                    .user(username)
                    .subscribeOn(Schedulers.newThread())
                    .observeOn(AndroidSchedulers.mainThread())
            )
    }

    private fun getRetrofit(): Service {
        if (retrofit == null) {
            retrofit =  RestAdapter.Builder()
                    .setEndpoint("https://api.github.com/")
                    .setClient(OkClient(client()))
                    .build()
        }
        if (service == null) {
            service = retrofit!!.create(Service::class.java)
        }
        return service!!
    }

    private fun client(): OkHttpClient {
        val client = OkHttpClient()
        client.setConnectTimeout(200, TimeUnit.SECONDS)
        client.setReadTimeout(200, TimeUnit.SECONDS)
        client.interceptors().add(Interceptor { chain ->
            val request = chain.request()
            val response: Response? = chain.proceed(request)
            Log.i("ServiceInfo", response?.request()?.url()?.toString())
            response
        })
        return client
    }

    private interface Service {

        @GET("/users/{user}")
        fun user(@Path("user") user: String): Observable<User>
    }

}
